cardano-api/src/Cardano/Api/MetaData.hs

```diff
@@ -0,0 +1,106 @@
+  deriving (Eq, Ord, Show)
+
+
+renderMetaDataError :: MetaDataError -> Text
+renderMetaDataError err =
+  case err of
+    ConversionErrBoolNotAllowed -> "JSON Bool value is not allowed in MetaData"
+    ConversionErrNullNotAllowed -> "JSON Null value is not allowed in MetaData"
+    ConversionErrNumberNotInteger -> "Only integers are allowed in MetaData"
+    ConversionErrLongerThan64Bytes -> "JSON string is longer than 64 bytes"
+    DecodeErrJSON fp decErr ->
+      "Error decoding JSON at: " <> textShow fp <> " Error: " <> textShow decErr
+    ReadFileErr fp rErr ->
+      "Error reading file at: " <> textShow fp <> " Error: " <> textShow rErr
+
+-- No 'ToJSON' instance is written for 'MetaData'.
+-- It does not make sense to JSON roundtrip test 'MetaData'
+-- as 'String's and 'ByteString's are indistinguishable.
+instance FromJSON MetaData where
+  parseJSON v = MetaData <$> parseJSON v
+
+instance FromJSON MetaDatum where
+  parseJSON v =
+    case valueToMetaDatum v of
+      Right mDatum -> return mDatum
+      Left decErr -> fail . show $ renderMetaDataError decErr
+
+readJSONMetaData :: FilePath -> IO (Either MetaDataError MetaData)
+readJSONMetaData fp = do
+  eBs <- Exception.try $ LC8.readFile fp
+  case eBs of
+    Left ioEx -> return . Left . ReadFileErr fp $ handler ioEx
+    Right bs -> return . first (DecodeErrJSON fp) $ eitherDecode' bs
+ where
+  handler :: IOException -> String
+  handler e = "Cardano.Api.MetaData.readJSONMetaData: " <> displayException e
+
+
+valueToMetaDatum :: Value -> Either MetaDataError MetaDatum
+valueToMetaDatum Null = Left ConversionErrNullNotAllowed
+valueToMetaDatum (Bool _) = Left ConversionErrBoolNotAllowed
+valueToMetaDatum (Number sci) =
+  case (floatingOrInteger sci :: Either Double Integer) of
+    Left _ -> Left ConversionErrNumberNotInteger
+    Right int -> Right $ I int
+valueToMetaDatum (String txt) = byteStringOrText txt
+valueToMetaDatum (Array vector) =
+  case Vector.mapM valueToMetaDatum vector of
+    Left err -> Left err
+    Right vecDatums -> Right . List $ Vector.toList vecDatums
+valueToMetaDatum (Object hm) =
+  case mapM tupleToMetaDatum $ HMS.toList hm of
+    Left err -> Left err
+    Right tuples -> Right $ Map tuples
+
+-- Helpers
+
+tupleToMetaDatum :: (Text, Value) -> Either MetaDataError (MetaDatum, MetaDatum)
+tupleToMetaDatum (k, v) = do
+  metaKey <- valueToMetaDatum $ String k
+  metaValue <- valueToMetaDatum v
+  Right (metaKey, metaValue)
+
+-- | If text is encoded in base64, we convert it to a 'ByteString'.
+byteStringOrText :: Text -> Either MetaDataError MetaDatum
+byteStringOrText txt
+  | SC8.length utf8 > 64 = Left ConversionErrLongerThan64Bytes
+  | B64.isBase64 utf8 = let bs = B64.encodeBase64' utf8
+                        in if SC8.length bs > 64
+                           then Left ConversionErrLongerThan64Bytes
+                           else Right $ B bs
+  | otherwise = Right $ S txt
+ where
+  utf8 = encodeUtf8 txt
```

Comment:
I think we need to swap these two clauses. If it's base64 we need to check that first. Otherwise valid base64 will also be utf8, but the base64 encoding as utf8 may be longer than 64, since base64 expands by 25%.

Also, as a general rule in Haskell we should prefer operations that "do it and tell you if it worked" over a pair of "function to test if it will work" plus "partial function that only works if you checked it first". For example, we avoid using List.null + List.head/List.tail in favour of pattern matching to see if it's [] or (:).

Here, we should not `B64.isBase64` and then `B64.encodeBase64'`. Rather we should just use a single `B64.decodeBase64` that simple tries it and tells you if it was valid, and if so gives the result. You can use this with pattern guards:
```
 byteStringOrText txt 
  | Just bs <- decodeBase64 txt = if if SC8.length bs > 64 then ... else ...
  | SC8.length utf8 > 64 = ...
  | otherwise = ...
```

Note also you were using `encodeBase64` and not `decodeBase64`.

http://hackage.haskell.org/package/base64-0.4.2/docs/Data-ByteString-Base64.html#v:decodeBase64